src/charm.py

```diff
@@ -2,17 +2,102 @@
 """Charm code for `mongos` daemon."""
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
+from typing import List
+from charms.mongodb.v1.helpers import copy_licenses_to_unit
+from charms.operator_libs_linux.v1 import snap
 
+from config import Config
+
+import ops
+from ops.model import (
+    BlockedStatus,
+    MaintenanceStatus,
+)
+from ops.charm import (
+    InstallEvent,
+    StartEvent,
+)
 
 import logging
 
-import ops
 
 logger = logging.getLogger(__name__)
 
+APP_SCOPE = Config.Relations.APP_SCOPE
+UNIT_SCOPE = Config.Relations.UNIT_SCOPE
+
 
 class MongosOperatorCharm(ops.CharmBase):
     """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.start, self._on_start)
+
+        # todo future PRs:
+        # 1. start daemon when relation to config server is made
+        # 2. add users for related application
+        # 3. update status indicates missing relations
+
+    # BEGIN: hook functions
+    def _on_install(self, event: InstallEvent) -> None:
+        """Handle the install event (fired on startup)."""
+        self.unit.status = MaintenanceStatus("installing mongos")
+        try:
+            self._install_snap_packages(packages=Config.SNAP_PACKAGES)
+
+        except snap.SnapError:
+            self.unit.status = BlockedStatus("couldn't install mongos")
+            return
+
+        # clear the default config file - user provided config files will be added in the config
+        # changed hook
+        try:
+            with open(Config.MONGOD_CONF_FILE_PATH, "r+") as f:
+                f.truncate(0)
+        except IOError:
+            self.unit.status = BlockedStatus("Could not install mongos")
```

Comment:
Maybe we can update the message to a user, because here we are failing on opening file, not on installing mongos, so the message is a bit misleading